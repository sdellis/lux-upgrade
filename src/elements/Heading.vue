<template>
  <component :is="level" class="lux-heading" :class="{'lux-hidden': hidden}">
    <slot/>
  </component>
</template>

<script>
/**
 * Headings are used as the titles of each major section of a page in the
 * interface. For example, templates generally use headings as their title.
 * Heading element provides an option to change the level of the heading.
 */
export default {
  name: "Heading",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  props: {
    /**
     * The heading level used for the heading.
     * `h1, h2, h3, h4, h5, h6`
     */
    level: {
      type: String,
      default: "h1",
      validator: value => {
        return value.match(/(h1|h2|h3|h4|h5|h6)/)
      },
    },
    /**
     * Whether the heading is visually hidden or not.
     * `true, false`
     */
    hidden: {
      type: Boolean,
      default: false,
    },
  },
}
</script>

<style lang="scss" scoped>
.lux-heading {
  @include reset;
  @include stack-space($space-base);
  font-family: $font-family-heading;
  line-height: $line-height-heading;
  color: $color-rich-black;
}

.lux-hidden {
  @include visually-hidden;
}

h1 {
  @include responsive-font(3vw, $font-size-xxx-large-min, $font-size-xxx-large-max, $font-size-xxx-large);
  letter-spacing: $letter-spacing-x-small;
  font-weight: $font-weight-bold;
}

h2 {
  @include responsive-font(2.5vw, $font-size-xx-large-min, $font-size-xx-large-max, $font-size-xx-large);
  letter-spacing: $letter-spacing-small;
  font-weight: $font-weight-bold;
}

h3 {
  @include responsive-font(2vw, $font-size-x-large-min, $font-size-x-large-max, $font-size-x-large);
  font-weight: $font-weight-bold;
}

h4 {
  @include responsive-font(1.5vw, $font-size-large-min, $font-size-large-max, $font-size-large);
  font-weight: $font-weight-semi-bold;
}

h5 {
  @include responsive-font(1vw, $font-size-base-min, $font-size-base-max, $font-size-base);
  font-weight: $font-weight-regular;
}

h6 {
  font-size: $font-size-small;
  font-weight: $font-weight-regular;
}
</style>

<docs>
  ```jsx
  <div>
    <heading level="h1" hidden>The quick brown fox</heading>
    <heading level="h2">The quick brown fox</heading>
    <heading level="h3">The quick brown fox</heading>
    <heading level="h4">The quick brown fox</heading>
  </div>
  ```

  ```jsx noeditor
  <div class="dos-n-donts">
    <div class="do">
      <div class="do-dont-example">
        <heading level="h1">Do this.</heading>
        <heading level="h2">Do this.</heading>
      </div>
      <p>Nest headings appropriately by level.</p>
    </div>

    <div class="dont">
      <div class="do-dont-example">
        <heading level="h2">Don't do this.</heading>
        <heading level="h1">Don't do this.</heading>
      </div>
      <p>If you nest headings out of order or skip levels, the lack of organization of content on the page will confuse users.</p>
    </div>
  </div>
  ```

</docs>
