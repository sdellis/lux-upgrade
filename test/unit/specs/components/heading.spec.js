import { createLocalVue, mount } from "@vue/test-utils"
import Heading from "@/elements/Heading.vue"

// create an extended `Vue` constructor
const localVue = createLocalVue()

describe("Heading.vue", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(Heading, {
      localVue,
      slots: {
        default: "Here's a heading.",
      },
    })
  })

  it("should render the appropriate level heading", () => {
    const heading1 = wrapper.find("h1")
    expect(heading1.is("h1")).toBe(true)
    wrapper.setProps({ level: "h3" })
    const heading3 = wrapper.find("h3")
    expect(heading3.is("h3")).toBe(true)
    expect(heading3.text()).toBe("Here's a heading.")
  })

  it("should display the slotted text", () => {
    wrapper.setProps({ level: "h2" })
    const heading2 = wrapper.find("h2")
    expect(heading2.is("h2")).toBe(true)
    expect(heading2.text()).toBe("Here's a heading.")
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
