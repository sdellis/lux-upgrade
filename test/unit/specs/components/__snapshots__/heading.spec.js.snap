// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading.vue has the expected html structure 1`] = `
<h1
  class="lux-heading"
>
  Here's a heading.
</h1>
`;
